plugins/chat/lib/chat_message_creator.rb

```diff
@@ -81,6 +86,34 @@ def validate_channel_status!
     end
   end
 
+  def validate_reply_chain!
+    return if @in_reply_to_id.blank?
+
+    @root_message_id = DB.query_single(<<~SQL).last
+      WITH RECURSIVE root_message_finder( id, in_reply_to_id )
+      AS (
+        -- start with the message id we want to find the parents of
+        SELECT id, in_reply_to_id
+        FROM chat_messages
+        WHERE id = #{@in_reply_to_id}
+
+        UNION ALL
+
+        -- get all parents of the message
+        SELECT cm.id, cm.in_reply_to_id
+        FROM root_message_finder rm
+        JOIN chat_messages cm ON rm.in_reply_to_id = cm.id
+      )
+      SELECT id FROM root_message_finder
+      WHERE in_reply_to_id IS NULL;
+    SQL
```

Comment:
Im starting to wonder if don't consider the possibility of having a `thread_original_message_id`

```
id: 1, in_reply_to: null, thread_id: 999, thread_original_message_id: null
id: 2, in_reply_to: 1, thread_id: 999, thread_original_message_id: 1
id: 3, in_reply_to: 2, thread_id: 999, thread_original_message_id: 1
```

Reply:
I don't think this is necessary since you can just get that information from the thread. However, I think maybe we are missing another scenario. If we are replying directly to a message, and that message already has a thread ID, we shouldn't have to do a root lookup since we can just assume that it's the correct thread.

Then we only need to do a root message lookup if there is no thread ID on the message we are replying to.

The other thing we can also do is pass the thread ID in the POST request if we already had it. I was kind of wary of this at first, but I think if we have proper validation and permissions checks it will be fine.

Reply:
I just don't think this is necessary at this point in time, will keep in mind though and add later if necessary, it will be easy to backfill.